sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1290,25 +1291,19 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
     }
   }
 
-  if (Device->useImmediateCommandLists()) {
-    auto &ComputeQueueGroupRef = ComputeQueueGroup;
-    bool ThreadSpecificImmCmdLists =
-        Device->useImmediateCommandLists() == PerThreadPerQueue;
-    if (ThreadSpecificImmCmdLists) {
-      // Thread id will be used to create separate imm cmdlists per thread.
-      auto TID = std::this_thread::get_id();
-      std::pair<std::unordered_map<std::thread::id, pi_queue_group_t>::iterator,
-                bool>
-          Result = ComputeQueueGroupsByTID.insert({TID, ComputeQueueGroup});
-      ComputeQueueGroupRef = Result.first->second;
-    }
-    // Create space to hold immediate commandlists corresponding to the
-    // ZeQueues
-    ComputeQueueGroupRef.ImmCmdLists = std::vector<pi_command_list_ptr_t>(
-        ComputeQueueGroup.ZeQueues.size(), CommandListMap.end());
-  }
+  // Thread id will be used to create separate queue groups per thread.
+  auto TID = std::this_thread::get_id();
```

Comment:
what if user requested NOT to create lists per-thread ?

Reply:
A single entry in the map is used. All threads use the same immediate commandlists.

Reply:
what makes it "single"? what if there are multiple threads in the app, they would have different thread IDs, and this multiple entries in the map, no?

Reply:
See function _pi_queue::getQueueGroup around line 1544. When per-thread is not requested then the first entry in the map is used, without looking at TID. That first entry is created at queue construction time.

Reply:
But why create multiple entries here if you are only going to use one?

Reply:
Multiple entries are not created:

  if (Device->useImmediateCommandLists() == PerThreadPerQueue) {

only if per-thread is requested do we create a new map entry. Otherwise, just the first entry is used.

Reply:
but a new map entry is created here: https://github.com/intel/llvm/pull/7041/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R1240

Reply:
There are two calls to "insert". The one in the _pi_queue constructor is called once at queue construction to create an entry using the TID of the thread creating the queue. The only other call to "insert" is in the getQueueGroup function, and that insert is under the if-condition that checks if per-thread-per-queue is set. If per-thread-per-queue is not set, then the map entry created at queue construction is used.

Reply:
I see now, thanks. It'd be good to describe this organization of the data structure in the header